### PR TITLE
[Arrow] Support consuming an "arrow_array_stream" PyCapsule

### PR DESCRIPTION
**Pull request by Tishj**
This PR implements https://github.com/duckdb/duckdb/discussions/10716

This allows consuming arrow array streams from PyCapsules, which do not necessarily have to be created by `pyarrow`.

For regular arrow scans we use `pyarrow` to perform filter pushdown on the stream, constructing a scanner to do so.
When scanning from a PyCapsule we disable this so no interaction with `pyarrow` is required.

### CODE REVIEW

File: tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py [hunk 1]
```diff
@@ -27,9 +21,24 @@ def test_arrow_table_replacement_scan(self, duckdb_cursor):
...
+
+        capsule = tbl.__arrow_c_stream__()
+        rel = duckdb_cursor.sql("select * from capsule where a > 3 and a < 5")
+        assert rel.fetchall() == [(4,)]
+
+        tbl = pa.Table.from_pydict({'a': [1, 2, 3], 'b': [4, 5, 6], 'c': [7, 8, 9], 'd': [10, 11, 12]})
+        capsule = tbl.__arrow_c_stream__()
+
+        rel = duckdb_cursor.sql("select b, d from capsule")
+        assert rel.fetchall() == [(i, i + 6) for i in range(4, 7)]
```
Comment (pdet) [inline tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py:39]: We could theoretically run this test without having any pyarrow dependency, right?
Maybe we add one for that

File: tools/pythonpkg/src/python_replacement_scan.cpp [hunk 2]
```diff
@@ -24,7 +25,13 @@ static void CreateArrowScan(const string &name, py::object entry, TableFunctionR
 	children.push_back(make_uniq<ConstantExpression>(Value::POINTER(CastPointerToValue(stream_factory_produce))));
 	children.push_back(make_uniq<ConstantExpression>(Value::POINTER(CastPointerToValue(stream_factory_get_schema))));
 
-	table_function.function = make_uniq<FunctionExpression>("arrow_scan", std::move(children));
+	if (type == PyArrowObjectType::PyCapsule) {
+		// Disable projection+filter pushdown
+		table_function.function = make_uniq<FunctionExpression>("arrow_scan_dumb", std::move(children));
```
Comment (pdet) [inline tools/pythonpkg/src/python_replacement_scan.cpp:30]: Maybe `arrow_no_pushdown`?

File: tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py [hunk 3]
```diff
@@ -3,19 +3,13 @@
...
-
-    can_run = True
-except:
-    can_run = False
+pa = pytest.importorskip("pyarrow")
+pq = pytest.importorskip("pyarrow.parquet")
+ds = pytest.importorskip("pyarrow.dataset")
 
 
 class TestArrowReplacementScan(object):
```
Comment (pdet) [inline tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py:11]: Could we also add test to trigger the exceptions added?
e.g., throw InvalidInputException("Expected a 'arrow_array_stream' PyCapsule, got: %s", string(capsule.name()));

**Review by pdet — CHANGES_REQUESTED** [review]
Hey @Tishj, thanks for picking up this improvement. It is looking good and I've added some minor comments.

File: tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py (hunk 1 above)
Comment (Tishj) [reply tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py:39]: We could, but I don't know if there's a library that produces an arrow pycapsule that does not depend on pyarrow

File: tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py (hunk 3 above)
Comment (Tishj) [reply tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py:11]: I'll add a test for that 👍

File: tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py (hunk 1 above)
Comment (Tishj) [reply tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py:39]: I think they do, but I thought Polars still has a dependency on pyarrow?
Might be entirely wrong on that actually, I'll check if I can install+run polars with pyarrow uninstalled

Comment (Tishj) [reply tools/pythonpkg/tests/fast/arrow/test_arrow_replacement_scan.py:39]: Added the test, they do not rely on pyarrow 👍

File: tools/pythonpkg/src/python_replacement_scan.cpp (hunk 2 above)
Comment (Tishj) [reply tools/pythonpkg/src/python_replacement_scan.cpp:30]: Hmm I agree, but it's used by duckdb-java, see the [initial PR](https://github.com/duckdb/duckdb/pull/5449) and the [JDBC Client source code](https://github.com/duckdb/duckdb-java/blob/2dca81c771e5af1db4074d9d7c0e7ab6f5a31ae3/src/jni/duckdb_java.cpp#L1235)

I'm hesitant to rename something that's used by third party clients

**Review by pdet — APPROVED** [review]
LGTM!

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**